_SPIKE_.md

```diff
@@ -0,0 +1,549 @@
+- MongoDB adds extra column-level option `text` as an alternative to ASC/DESC.
+- PostgreSQL allows column expressions, e.g. `lower(email)`.
+
+Problem 2: do we want to use database column names or LB property names? These
+two names can be different, especially when using database-specific config to
+map property names to custom column names.
+
+**The proposal:**
+
+At high-level, keep the current syntax where indexes are defined via a key-value
+map stored in `settings.indexes` property, the key is the index name and the
+value is an index definition object.
+
+```ts
+@model({
+  strict: false,
+  forceID: true,
+  indexes: {
+    uniqueEmail: {
+      // index definition
+    },
+    nameQueries: {
+      // index definition
+    },
+  },
+})
+class MyModel extends Entity {}
+```
+
+Individual indexes can be defined as follows:
+
+- Add a new field `properties` as a key-value map from property names to
+  indexing order:
+
+  ```js
+  // definition of an individual index
+  emailIndex: {
+    properties: {
+      email: 1, // ASC
+      createdAt: 'DESC', // alias for -1
+      bio: 'text', // database-specific value (MongoDB's "text")
+    }
+  }
+  ```
+
+  Important: property names are mapped to database column names when building
+  the index definition.
+
+- Keep supporting `keys` field as a key-value map from database column name to
+  indexing order, see the description of the actual status below. Entries from
+  `keys` should be merged with entries from `properties`, `keys` taking
+  precedence (replacing `properties` entries).
+
+- Keep supporting `unique` field (set it to `true` to let the index enforce
+  uniqueness).
+
+- Database-specific options will be stored under a key with the connector name:
+
+  ```js
+  emailIndex: {
+    properties: {
+      email: 'ASC',
+    },
+    mongodb: {
+      sparse: true,
+    },
+    mysql: {
+      kind: 'fulltext',
+      type: 'hash',
+    },
+    postgresql: {
+      type: 'hash',
+    }
+  }
+  ```
+
+- To support more complex indexes offered by e.g. PostgreSQL, we should support
+  `columns` field too. This field should contain an array of column-index
+  specification, for example:
+
+  ```js
+  ['email ASC', 'lower(name) DESC'];
+  ```
+
+  In the contrary with the current implementation, I am proposing that `columns`
+  should replace any column list created from `properties` and `keys`, and that
+  `columns` should be nested inside connector-specific options:
+
+  ```js
+  emailIndex: {
+    properties: {
+      // default setup used by most connectors
+      email: 'ASC',
+    },
+    postgresql: {
+      // in PostgreSQL, use a different index definition
+      columns: ['lower(email) ASC']
+    },
+  }
+  ```
+
+### Foreign keys at property level
+
+Introduce a new property metadata "references" (inspired by ANSI SQL):
+
+```ts
+@property({
+  type: 'number',
+  required: true,
+  references: {
+    // a TypeResolver
+    model: () => Category,
```

Comment:
I'm wondering why we'd like to support three ways of declaring the target model here. Can we keep this simple and support the typeResolver flavour only (at least for first iteration)?

Reply:
Good point :+1: